toolkit/pallets/block-production-log/src/lib.rs

```diff
@@ -96,5 +112,15 @@ pub mod pallet {
 			Log::<T>::put(to_retain);
 			to_return
 		}
+
+		pub fn peek_prefix<'a>(slot: Slot) -> impl Iterator<Item = (Slot, T::BlockProducerId)> {
+			Log::<T>::get().into_iter().filter(move |(s, _)| s <= &slot)
+		}
+
+		pub fn drop_prefix<'a>(slot: Slot) {
+			let entries_left: Vec<_> =
+				Log::<T>::get().into_iter().filter(move |(s, _)| s > &slot).collect();
+			Log::<T>::put(entries_left);
+		}
```

Comment:
we can optimize it slightly, so try to mutate it and use retain directly on vector instead of call on iterator and collect (please double check due I have not tested it)
```suggestion
		pub fn drop_prefix<'a>(slot: &Slot) {
			Log::<T>::mutate(|log| log.retain(|(s, _)| s > slot));
		}
```

Reply:
that worked